### DISCUSSION TITLE
Can I create a reactive x-data variable in the directive itself?

### DISCUSSION THREAD

**Discussion by leobm** [alpinejs/alpine#2778] [1. Help] [answered]
Hi,

So a variable that is not previously known in x-data?
Please look at my example.
In the first x-data there is an x variable, in the second x-data there is not. 
But the second nam2 X: span is only initialized and then not updated.

Thanks, Felix

```html 
<!DOCTYPE html>
<html lang="en">
<head>
    <script src="//unpkg.com/alpinejs" defer></script>
    <script>
        document.addEventListener('alpine:init', () => {

            Alpine.directive('test', (el, {value, modifiers, expression}, {effect, evaluate, evaluateLater}) => {
                if (el.hasAttribute("x-model")) {
                    let modelExpression = el.getAttribute("x-model");
                    let getThing = evaluateLater(modelExpression);
                    effect(() => {
                        getThing(thing => {
                            evaluate("x=_rand", {scope: {_rand: Math.random()}});
                        });
                    });
                }
            });
        });
    </script>
</head>
<body>

<div x-data="{name: 'name1', x:0}">
    <input x-model="name" x-test/>
    <p>Name: <span x-text="name"></span></p>
    <p>X: <span x-text="x"></span></p>
</div>

<div x-data="{name: 'name2'}">
    <input x-model="name" x-test/>
    <p>Name: <span x-text="name"></span></p>
    <p>X: <span x-text="x"></span></p>
</div>

</body>
</html>
```

**Comment by SimoTod** [comment, marked as answer]
It's a bit more complex, the reactivity is set by the effect on the x-text directive, since x is not defined at that point, the effect won't be registered (kind of, it is to make it simpler). It's also a weird practice to use a variable that doesn't exist, you could define it with a sensible default (empty string, null, undefined or whatever you think it's the best for your use case). If you don't define it first, the answer is no, your existing directives won't react to the new variable.

**Reply by leobm** [reply]
ok, makes sense. I solve it then differently, you are certainly right, should define a default value before. 

thank you...
